Sources/NIO/ControlMessage.swift

```diff
@@ -0,0 +1,245 @@
+}
+
+/// Extract information from a collection of control messages.
+struct ControlMessageParser {
+    var ecnValue: NIOExplicitCongestionNotificationState = .transportNotCapable // Default
+    
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    private static let ipv4TosType = IP_RECVTOS
+    #else
+    private static let ipv4TosType = IP_TOS    // Linux
+    #endif
+
+    static func readCInt(data: UnsafeRawBufferPointer) -> CInt {
+        assert(data.count == MemoryLayout<CInt>.size)
+        precondition(data.count >= MemoryLayout<CInt>.size)
+        var readValue = CInt(0)
+        withUnsafeMutableBytes(of: &readValue) { valuePtr in
+            valuePtr.copyMemory(from: data)
+        }
+        return readValue
+    }
+    
+    mutating func receiveMessage(_ controlMessage: UnsafeControlMessage) {
+        if controlMessage.level == IPPROTO_IP && controlMessage.type == ControlMessageParser.ipv4TosType {
+            if let data = controlMessage.data {
+                assert(data.count == 1)
+                precondition(data.count >= 1)
+                let readValue = CInt(data[0])
+                self.ecnValue = .init(receivedValue: readValue)
+            }
+        } else if controlMessage.level == IPPROTO_IPV6 && controlMessage.type == IPV6_TCLASS {
+            if let data = controlMessage.data {
+                let readValue = ControlMessageParser.readCInt(data: data)
+                self.ecnValue = .init(receivedValue: readValue)
+            }
+        }
+    }
+}
+
+extension NIOExplicitCongestionNotificationState {
+    /// Initialise a NIOExplicitCongestionNotificationState from a value received via either TCLASS or TOS cmsg.
+    init(receivedValue: CInt) {
+        switch receivedValue & IPTOS_ECN_MASK {
+        case IPTOS_ECN_ECT1:
+            self = .transportCapableFlag1
+        case IPTOS_ECN_ECT0:
+            self = .transportCapableFlag0
+        case IPTOS_ECN_CE:
+            self = .congestionExperienced
+        default:
+            self = .transportNotCapable
+        }
+    }
+}
+
+extension CInt {
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    fileprivate static let notCapableValue = IPTOS_ECN_NOTECT
+    #else
+    fileprivate static let notCapableValue = IPTOS_ECN_NOT_ECT    // Linux
+    #endif
+
+    /// Create a CInt encoding of ExplicitCongestionNotification suitable for sending in TCLASS or TOS cmsg.
+    init(ecnValue: NIOExplicitCongestionNotificationState) {
+        switch ecnValue {
+        case .transportNotCapable:
+            self = CInt.notCapableValue
+        case .transportCapableFlag0:
+            self = IPTOS_ECN_ECT0
+        case .transportCapableFlag1:
+            self = IPTOS_ECN_ECT1
+        case .congestionExperienced:
+            self = IPTOS_ECN_CE
+        }
+    }
+}
+
+struct UnsafeOutboundControlBytes {
+    private var controlBytes: UnsafeMutableRawBufferPointer
+    private var writePosition: UnsafeMutableRawBufferPointer.Index = 0
```

Comment:
For the sake of keeping this as correct as possible, let's initialise this in `init` with `self.controlBytes.startIndex`.